src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferDictionaryContainsMethods.cs

```diff
@@ -0,0 +1,223 @@
+                //      3) Its name is "ContainsKey" if property is "Keys", or "ContainsValue" if property is "Values". 
+                //      4) It has exactly one parameter of the correct type:
+                //          - If the method is a "ContainsKey" method, its type must be TKey. 
+                //          - If the method is a "ContainsValue" method, its type must be TValue.
+
+                var invocation = (IInvocationOperation)context.Operation;
+                IMethodSymbol containsMethod = invocation.TargetMethod;
+
+                if (containsMethod.Name != ContainsMethodName)
+                    return;
+                if (containsMethod.ReturnType.SpecialType != SpecialType.System_Boolean)
+                    return;
+                if (!TryGetPropertyReferenceOperation(invocation, out IPropertySymbol? property))
+                    return;
+                if (!DoesContainsCandidateHaveCorrectArgumentCount(containsMethod))
+                    return;
+                if (!TryGetConstructedDictionaryType(property.ContainingType, out INamedTypeSymbol? constructedDictionaryType))
+                    return;
+
+                //  At this point, we know that the method being invoked is a method called "Contains" that has a boolean return type.
+                //  We also know that the method is being invoked on a property belonging to a type that implements IDictionary`2. We will
+                //  compare the types used to construct IDictionary`2 to the parameter type of the Contains method. 
+
+                ITypeSymbol keyType = constructedDictionaryType.TypeArguments[0];
+                ITypeSymbol valueType = constructedDictionaryType.TypeArguments[1];
+
+                //  We use Parameters.Last() because the first argument could be the key/value collection, depending
+                //  on whether the method is an extension method and whether the language is C# or Visual Basic. 
+
+                ITypeSymbol containsParameterType = containsMethod.Parameters.Last().Type;
+
+                if (property.Name == KeysPropertyName)
+                {
+                    if (!containsParameterType.Equals(keyType, SymbolEqualityComparer.Default))
+                        return;
+
+                    //  Now we search for an accessible ContainsKey method that returns boolean and accepts a single
+                    //  parameter of the type that was substituted for TKey.
+
+                    IMethodSymbol? containsKeyMethod = property.ContainingType.GetMembers(ContainsKeyMethodName)
+                        .OfType<IMethodSymbol>()
+                        .WhereAsArray(x => x.ReturnType.SpecialType == SpecialType.System_Boolean &&
+                            x.IsPublic() &&
+                            x.Parameters.Length == 1 &&
+                            x.Parameters[0].Type.Equals(keyType, SymbolEqualityComparer.Default))
+                        .SingleOrDefault();
```

Comment:
I understand the use of `SingleOrDefault` because we don't really expect to find multiple methods but I am always a bit reluctant to have failing analyzers. Let's wait for other inputs here.

Reply:
Now that I think about it, I think I actually did run into issues with `SingleOrDefault` on another PR when analyzing broken code. I'll change this to `FirstOrDefault`.